### DISCUSSION TITLE
Problem CTKToplevel - How i can keep focus second instance of window.

### DISCUSSION THREAD

**Discussion by CeContreras** [TomSchimansky/CustomTkinter#1689] [General]
When i want create new instance of CTkToplevel, it is created below of the main window. I dont know how i can keep the focus of second instance window. 
Pleace, who can help me?

**Comment by avalon60** [comment]


**Reply by CeContreras** [reply]
Could you tell me where I should set it in the code?

class ToplevelWindow(customtkinter.CTkToplevel):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.geometry("400x300")

        self.label = customtkinter.CTkLabel(self, text="ToplevelWindow")
        self.label.pack(padx=20, pady=20)


class App(customtkinter.CTk):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.geometry("500x400")

        self.button_1 = customtkinter.CTkButton(self, text="open toplevel", command=self.open_toplevel)
        self.button_1.pack(side="top", padx=20, pady=20)

        self.toplevel_window = None

    def open_toplevel(self):
        if self.toplevel_window is None or not self.toplevel_window.winfo_exists():
            self.toplevel_window = ToplevelWindow(self)  # create window if its None or destroyed
        else:
            self.toplevel_window.focus()  # if window exists focus it


app = App()
app.mainloop()

**Reply by avalon60** [reply]
I'm not sure why you have coded it this way, making the ToplevelWindow class, a subclass of CTkToplevel, which  essentially adds nothing to the new class. Also you need to block your code properly. I'm not sure whether it's the incorrect  block markdown, but does this code even do anything? You have either defined the `__init__() `method incorrectly,  or it looks incorrect, possibly due to your code being gashed because of bad markdown. 

Please correct so that I can get a better picture.

**Reply by CeContreras** [reply]
[![image](https://github.com/TomSchimansky/CustomTkinter/assets/42829956/e6758f7d-5bd3-4c80-911b-3614746165be)](https://customtkinter.tomschimansky.com/documentation/windows/toplevel) 

The code is described in the official customtkinter documentation, however when applied to a real project, the CTkToplevel instance does not maintain focus.

**Reply by avalon60** [reply]
It's 1:05am here now, I'll take a look in the morning.

**Reply by CeContreras** [reply]
thank you very much, stay tuned

**Reply by avalon60** [reply]
OK. I'm at the point where I am guessing what you are aiming for (but I think it's a good guess). Literally speaking the code does what you are saying you want it to do, although it may not be that obvious. It does maintain focus on the CTkTopLevel component, however, depending on where the windows are positioned, if you click the button, the CTkTopLevel can become hidden behind the root window. It still has focus, but it is hidden. So I am guessing what you want, is for it to stay in front of the root window, in which case:

```
import customtkinter

class ToplevelWindow(customtkinter.CTkToplevel):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.geometry("400x300")

        self.label = customtkinter.CTkLabel(self, text="ToplevelWindow")
        self.label.pack(padx=20, pady=20)


class App(customtkinter.CTk):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.geometry("500x400")

        self.button_1 = customtkinter.CTkButton(self, text="open toplevel", command=self.open_toplevel)
        self.button_1.pack(side="top", padx=20, pady=20)

        self.toplevel_window = None

    def open_toplevel(self):
        if self.toplevel_window is None or not self.toplevel_window.winfo_exists():
            self.toplevel_window = ToplevelWindow(self)  # create window if its None or destroyed
        else:
            self.toplevel_window.lift()   # Lift CTkTopLevel window to the front 
            self.toplevel_window.focus()  # if window exists focus it

app = App()
app.mainloop()
```
So the` self.toplevel_window.lift()` (4th line from the bottom) has been added to achieve this.

There is another possibility, which you may be looking for, and that is to make the CTkTopLevel window modal, blocking input to the root window. You can do that with this mod:

```
class ToplevelWindow(customtkinter.CTkToplevel):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.geometry("400x300")

        self.label = customtkinter.CTkLabel(self, text="ToplevelWindow")
        self.label.pack(padx=20, pady=20)

        self.update_idletasks()
        self.grab_set()

```
The last 2 lines achieve this, making the CTkTopLevel the only window which will accept input, until it is closed.

Hope one of  these provides the answer that you are looking for.

**Reply by CeContreras** [reply]
This solution worked for me!

-self.update_idletasks()  
-self.grab_set() 

Last question, how can I destroy a button, since I'm using .destroy() and it doesn't work for me.

**Reply by avalon60** [reply]
> This solution worked for me!
> 
> -self.update_idletasks() -self.grab_set()
> 
> Last question, how can I destroy a button, since I'm using .destroy() and it doesn't work for me.

Where are you placing the destroy statement?

**Reply by CeContreras** [reply]
![image](https://github.com/TomSchimansky/CustomTkinter/assets/42829956/41818bd4-c4ee-4d39-ad25-f9dbe73932cc)
In the line 735

**Reply by CeContreras** [reply]
How do I change the CTkToplevel icon, when I try to change it with self.iconbitmap("images/ico.ico"), it doesn't allow and handles customtkinter's.

**Reply by avalon60** [reply]
I think there is currently a bug with window icons not working in CTkTopLevel.

Re destroy. Have you tried removing the try/except block to see if an exception is being raised?

Btw, it's considered bad practice to have a bare except, in try / except blocks. You should trap explicit conditions.

**Reply by avalon60** [reply]
Here is how you would destroy the button in the original sample code:

```
class ToplevelWindow(customtkinter.CTkToplevel):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        attrs = vars(self)
        # print(f'Attrs = {attrs}')
        self.geometry("400x300")
        print(self.master)
        self.master.button_1.destroy()

        self.label = customtkinter.CTkLabel(self, text="ToplevelWindow")
        self.label.pack(padx=20, pady=20)
        self.master.button_1.destroy()
```
The last line points to the button defined in it's parent object (CTk) and calls the destroy method.

**Reply by CeContreras** [reply]
![image](https://github.com/TomSchimansky/CustomTkinter/assets/42829956/fec5af69-4442-402a-956e-f14088209c21)
i listened to your instruccions, but not deleted the button

**Reply by CeContreras** [reply]
when you said that error with CTkToplevel icon, it's because from the framework the layout is like this?

**Reply by CeContreras** [reply]
I have a question, how can I make CTkToplevel show up in the center when initial?

**Reply by avalon60** [reply]
> when you said that error with CTkToplevel icon, it's because from the framework the layout is like this?

No error - it just doesn't render the icon.

**Reply by CeContreras** [reply]
Tengo una pregunta, ¿cómo puedo hacer que CTkToplevel aparezca en el centro cuando se inicia?

**Comment by Akascape** [comment]
@CeContreras This can be fixed using these methods:

```python
toplevel.attributes('-topmost', 1)
toplevel.attributes('-topmost', 0)
toplevel.focus()
```
Or
```python
toplevel.resizable(True,True)
toplevel.transient(root)
```
